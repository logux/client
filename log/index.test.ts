import { TestPair, TestTime, TestLog } from '@logux/core'
import { yellow, bold } from 'colorette'

import { CrossTabClient, ClientMeta, log } from '../index.js'

function setState (client: any, state: string) {
  client.node.setState(state)
}

function emit (obj: any, event: string, ...args: any[]) {
  obj.emitter.emit(event, ...args)
}

function privateMethods (obj: object): any {
  return obj
}

async function createClient () {
  let pair = new TestPair()
  privateMethods(pair.left).url = 'ws://example.com'
  let client = new CrossTabClient<{}, TestLog<ClientMeta>>({
    subprotocol: '1.0.0',
    server: pair.left,
    userId: '10',
    time: new TestTime()
  })

  client.role = 'leader'
  client.node.catch(() => {})

  await pair.left.connect()
  return client
}

let group = false
let out = ''

function format (...args: (string | object)[]): string {
  return (
    (group ? '  ' : '') +
    args
      .filter(i => {
        if (typeof i === 'string') {
          return i !== '' && !i.includes('font-weight:')
        } else {
          return true
        }
      })
      .map(i => {
        if (typeof i === 'string') {
          return i
            .replace(/%cLogux%c/, yellow(bold('Logux')))
            .replace(/%c([^%]+)(%c)?/g, bold('$1'))
        } else {
          return JSON.stringify(i)
        }
      })
      .join(' ')
  )
}

beforeAll(() => {
  jest.spyOn(console, 'groupCollapsed').mockImplementation((...args: any[]) => {
    console.log(...args)
    group = true
  })
  jest.spyOn(console, 'groupEnd').mockImplementation(() => {
    group = false
  })
  jest.spyOn(console, 'log').mockImplementation((...args: any[]) => {
    out += format(...args) + '\n'
  })
})

beforeEach(() => {
  out = ''
})

it('prints log', async () => {
  let client = await createClient()
  client.node.connected = false
  setState(client, 'disconnected')
  log(client)

  emit(client, 'role')
  setState(client, 'connecting')

  client.node.remoteNodeId = 'server:uuid'
  client.node.connected = true
  setState(client, 'synchronized')

  await client.node.log.add({ type: 'A' }, { sync: true })
  setState(client, 'sending')
  setState(client, 'synchronized')
  await client.node.log.add(
    {
      type: 'logux/undo',
      action: { type: 'A' },
      id: '1 10:1:1 0',
      reason: 'error'
    },
    { id: '1 server:uuid 0' }
  )

  await client.node.log.add(
    { type: 'logux/subscribe', channel: 'users' },
    { sync: true }
  )
  setState(client, 'sending')
  setState(client, 'synchronized')
  await client.node.log.add(
    { type: 'A' },
    { sync: true, id: '2 server:uuid 1' }
  )
  await client.node.log.add(
    {
      type: 'logux/processed',
      id: '2 10:1:1 0'
    },
    { id: '2 server:uuid 0' }
  )

  client.node.connected = false
  setState(client, 'disconnected')

  setState(client, 'connecting')
  client.node.connected = true
  setState(client, 'synchronized')

  await client.node.log.add(
    { type: 'logux/subscribe', channel: 'users', since: 1 },
    { sync: true }
  )
  setState(client, 'sending')
  setState(client, 'synchronized')
  await client.node.log.add(
    {
      type: 'logux/undo',
      action: { type: 'logux/subscribe', channel: 'users', since: 1 },
      id: '3 10:1:1 0',
      reason: 'error',
      wrongProp: 'sync'
    },
    { id: '3 server:uuid 0' }
  )
  await client.node.log.add(
    { type: 'logux/unsubscribe', channel: 'users' },
    { sync: true }
  )
  await client.node.log.add(
    {
      type: 'logux/unsubscribe',
      channel: 'users',
      bad: true
    },
    { sync: true }
  )

  await client.node.log.add({ type: 'B' }, { sync: true })
  setState(client, 'sending')
  setState(client, 'synchronized')
  await client.node.log.add(
    {
      type: 'logux/processed',
      id: '7 10:1:1 0'
    },
    { id: ' server:uuid 0' }
  )

  await client.node.log.add(
    {
      type: 'logux/processed',
      id: '200 10:1:1 0'
    },
    { id: '4 server:uuid 0' }
  )

  await client.node.log.add(
    {
      type: 'logux/undo',
      action: { type: 'B' },
      id: '201 10:1:1 0',
      reason: 'error'
    },
    { id: '5 server:uuid 0' }
  )

  client.nodeId = '20:2:2'
  emit(client, 'user', '20')

  expect(out).toMatchSnapshot()
})

it('returns unbind function', async () => {
  let client = await createClient()
  let unbind = log(client)

  unbind()
  emit(client, 'role')

  expect(out).toEqual('')
})

it('supports cross-tab synchronization', async () => {
  let client = await createClient()
  client.role = 'follower'
  log(client)

  emit(client, 'role')
  client.state = 'connecting'
  emit(client, 'state')

  expect(out).toMatchSnapshot()
})

it('ignores different tab actions', async () => {
  let client = await createClient()
  log(client)

  await client.node.log.add({ type: 'A' }, { tab: 'X', reasons: ['test'] })
  await client.node.log.removeReason('test')

  expect(out).toEqual('')
})

it('ignores actions by request', async () => {
  let client = await createClient()
  log(client, { ignoreActions: ['A', 'B'] })

  await Promise.all([
    client.node.log.add({ type: 'A' }, { reasons: ['test'] }),
    client.node.log.add({ type: 'B' })
  ])
  await client.node.log.removeReason('test')
  await client.node.log.add({ type: 'C' })

  expect(out).toMatchSnapshot()
})
