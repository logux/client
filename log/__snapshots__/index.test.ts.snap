// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ignores actions by request 1`] = `
"[33m[1mLogux[22m[39m added and cleaned [1mC[22m action
  Action {\\"type\\":\\"C\\"}
  Meta {\\"id\\":\\"3 10:1:1 0\\",\\"time\\":3,\\"reasons\\":[],\\"subprotocol\\":\\"1.0.0\\"}
"
`;

exports[`prints log 1`] = `
"[33m[1mLogux[22m[39m tab role is [1mleader[22m
[33m[1mLogux[22m[39m state is [1mconnecting[22m
  Node ID: [1m10:1:1[22m
  Server: [1mws://example.com[22m
[33m[1mLogux[22m[39m state is [1msynchronized[22m
  Server ID: [1mserver:uuid[22m
[33m[1mLogux[22m[39m added [1mA[22m action
  Action {\\"type\\":\\"A\\"}
  Meta {\\"sync\\":true,\\"id\\":\\"1 10:1:1 0\\",\\"time\\":1,\\"reasons\\":[\\"syncing\\"],\\"subprotocol\\":\\"1.0.0\\",\\"added\\":1}
[33m[1mLogux[22m[39m state is [1msending[22m
[33m[1mLogux[22m[39m state is [1msynchronized[22m
[33m[1mLogux[22m[39m cleaned [1mA[22m action
  Action {\\"type\\":\\"A\\"}
  Meta {\\"sync\\":true,\\"id\\":\\"1 10:1:1 0\\",\\"time\\":1,\\"reasons\\":[],\\"subprotocol\\":\\"1.0.0\\",\\"added\\":1}
[33m[1mLogux[22m[39m action [1mA[22m was undone because of [1merror[22m
  Reverted Action {\\"type\\":\\"A\\"}
[33m[1mLogux[22m[39m subscribing to [1musers[22m channel
[33m[1mLogux[22m[39m state is [1msending[22m
[33m[1mLogux[22m[39m state is [1msynchronized[22m
[33m[1mLogux[22m[39m added [1mA[22m action
  Action {\\"type\\":\\"A\\"}
  Meta {\\"sync\\":true,\\"id\\":\\"2 server:uuid 1\\",\\"time\\":2,\\"reasons\\":[\\"syncing\\"],\\"added\\":3}
  From: [1mserver:uuid[22m
[33m[1mLogux[22m[39m subscribed to [1musers[22m channel
  Processed Action {\\"type\\":\\"logux/subscribe\\",\\"channel\\":\\"users\\"}
[33m[1mLogux[22m[39m state is [1mdisconnected[22m
[33m[1mLogux[22m[39m state is [1mconnecting[22m
  Node ID: [1m10:1:1[22m
  Server: [1mws://example.com[22m
[33m[1mLogux[22m[39m subscribing to [1musers[22m channel
  Action {\\"type\\":\\"logux/subscribe\\",\\"channel\\":\\"users\\",\\"since\\":{\\"id\\":\\"2 server:uuid 0\\",\\"time\\":2}}
[33m[1mLogux[22m[39m state is [1msynchronized[22m
  Server ID: [1mserver:uuid[22m
[33m[1mLogux[22m[39m subscribing to [1musers[22m channel
  Action {\\"type\\":\\"logux/subscribe\\",\\"channel\\":\\"users\\",\\"since\\":1}
[33m[1mLogux[22m[39m state is [1msending[22m
[33m[1mLogux[22m[39m state is [1msynchronized[22m
[33m[1mLogux[22m[39m subscription to [1musers[22m was undone because of [1merror[22m
  Reverted Action {\\"type\\":\\"logux/subscribe\\",\\"channel\\":\\"users\\",\\"since\\":1}
  Undo Action {\\"type\\":\\"logux/undo\\",\\"action\\":{\\"type\\":\\"logux/subscribe\\",\\"channel\\":\\"users\\",\\"since\\":1},\\"id\\":\\"3 10:1:1 0\\",\\"reason\\":\\"error\\",\\"wrongProp\\":\\"sync\\"}
[33m[1mLogux[22m[39m unsubscribed from channel [1musers[22m
[33m[1mLogux[22m[39m unsubscribed from channel [1musers[22m
  Action {\\"type\\":\\"logux/unsubscribe\\",\\"channel\\":\\"users\\",\\"bad\\":true}
[33m[1mLogux[22m[39m added [1mB[22m action
  Action {\\"type\\":\\"B\\"}
  Meta {\\"sync\\":true,\\"id\\":\\"7 10:1:1 0\\",\\"time\\":7,\\"reasons\\":[\\"syncing\\"],\\"subprotocol\\":\\"1.0.0\\",\\"added\\":7}
[33m[1mLogux[22m[39m state is [1msending[22m
[33m[1mLogux[22m[39m state is [1msynchronized[22m
[33m[1mLogux[22m[39m cleaned [1mB[22m action
  Action {\\"type\\":\\"B\\"}
  Meta {\\"sync\\":true,\\"id\\":\\"7 10:1:1 0\\",\\"time\\":7,\\"reasons\\":[],\\"subprotocol\\":\\"1.0.0\\",\\"added\\":7}
[33m[1mLogux[22m[39m action [1mB[22m was processed
  Processed Action {\\"type\\":\\"B\\"}
[33m[1mLogux[22m[39m action [1m200 10:1:1 0[22m was processed
[33m[1mLogux[22m[39m action [1mB[22m was undone because of [1merror[22m
  Reverted Action {\\"type\\":\\"B\\"}
[33m[1mLogux[22m[39m user ID was changed to [1m20[22m
  Node ID: [1m20:2:2[22m
"
`;

exports[`supports cross-tab synchronization 1`] = `
"[33m[1mLogux[22m[39m tab role is [1mfollower[22m
[33m[1mLogux[22m[39m state is [1mconnecting[22m
  Node ID: [1m10:1:1[22m
  Server: [1mws://example.com[22m
"
`;
